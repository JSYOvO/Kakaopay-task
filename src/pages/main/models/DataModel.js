export default {
    getList() {
        return new Promise((resolve, reject) => {
            fetch('https://my-json-server.typicode.com/kakaopay-fe/resources/words')
            .then((res) => {
                resolve(res.json());
            })
            .catch((err) => {
                console.log("Fetch Error!!!", err);
            })
        })
    }
}