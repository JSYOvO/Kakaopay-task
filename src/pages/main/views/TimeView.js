import View from './View.js'
const TimeView = Object.create(View) // 게임의 남은시간을 사용자에게 전달하기 위한 뷰

TimeView.setup = function (el) {
    this.init(el) // html 요소 등록
    this.discountLeftTime = 0; // setInterval를 활용해 남은시간을 줄이기 위해 선언
    this.leftTimeEl = el.querySelector('.header__time__left'); // 게임의 남은시간 영역
    this.showLeftTime(0); // 0으로 남은시간 초기화
    this.leftTime = 0; // 현재 남은 시간
    this.initialTime = 0; // 각 단어의 제한시간
    return this
}

TimeView.renderGameData = function (time) { // 각 단어의 제한 시간 출력 및 매초 감소
  this.initialTime = this.leftTime = time;
  this.showLeftTime(this.leftTime); // 남은 시간 출력

  this.discountLeftTime = setInterval(() => { // setInterval을 활용하여 매초 남은시간 감소
    this.leftTime--; 
    this.showLeftTime(this.leftTime);
    
    if(this.leftTime == 0) { // 만일 남은시간이 0라면 setInterval 종료
      clearInterval(this.discountLeftTime);
      this.emit('@timeout'); // @timeout 커스텀 이벤트 생성
    }
  }, 1000)
}

TimeView.showLeftTime = function (leftTime) {
  this.leftTimeEl.innerHTML = leftTime + '초';
}

TimeView.renderInitialDate = function () { // 남은시간 0으로 초기화
  clearInterval(this.discountLeftTime); // 만약에 실행중 일수도 있는 setInterval 중지
  this.showLeftTime(0);
}

TimeView.getElapsedTime = function(initialTime = this.initialTime, leftTime = this.leftTime) { // 컨트롤러에서 문제를 푸는데 경과시간을 요청시 반환

  return new Promise(res => {
    res(initialTime - leftTime);
  })
}

export default TimeView;