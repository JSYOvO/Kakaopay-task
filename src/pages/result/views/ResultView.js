import View from './View.js'
const ResultView = Object.create(View) // 게임결과 사용자에게 전달하기 위한 뷰

ResultView.setup = function (el, totalScroe, avgTime) {
    this.init(el)
    this.resultTop = el.querySelector('h1'); // 게임 점수를 포함하는 영역
    this.resultDown = el.querySelector('p'); // 게임 평균 답변시간을 포함하는 영역
    this.resultTop.innerHTML = this.renderResultDataScore(totalScroe);
    this.resultDown.innerHTML = this.renderResultDataAvgTime(avgTime);
    // router push 호출시 같이 넘어온 totalScore, avgTime인자를 활용하여 게임 점수 및 평균 답변시간 화면에 출력
    return this;
}

ResultView.renderResultDataScore = function(totalScroe = 10) {
    return "당신의 점수는 " + totalScroe + "점입니다.";
}

ResultView.renderResultDataAvgTime = function(avgTime = 10) {
    return "단어당 평균 답변 시간은 " + parseInt(avgTime) + "초입니다.";
}

export default ResultView;