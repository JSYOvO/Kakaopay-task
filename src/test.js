import DataModel from './pages/main/models/DataModel';
import ResultView from './pages/result/views/ResultView';
import TimeView from './pages/main/views/TimeView';

test("[DataModel] fetch data test", () => {
    return DataModel.getList()
    .then(data => {
        expect(data).toMatchSnapshot();
        
        for(let i = 0; i< data.length; i += 1){
            expect(data[i]).toHaveProperty('second');
            expect(data[i]).toHaveProperty('text');
        }  
    })
});

test("[TimeView] getElapsedTime", () => {
    return TimeView.getElapsedTime(10, 5)
    .then(data => {
        expect(data).toEqual(5);
    })
})

test("[ResultView - renderResultDataScore]", () => {
    expect(ResultView.renderResultDataScore(5)).toEqual("당신의 점수는 5점입니다.");
    expect(ResultView.renderResultDataScore()).toEqual("당신의 점수는 10점입니다.");
});

test("[ResultView - renderResultDataAvgTime]", () => {
    expect(ResultView.renderResultDataAvgTime(5)).toEqual("단어당 평균 답변 시간은 5초입니다.");
    expect(ResultView.renderResultDataAvgTime()).toEqual("단어당 평균 답변 시간은 10초입니다.");
});