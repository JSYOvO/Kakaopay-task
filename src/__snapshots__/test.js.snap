// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[DataModel] fetch data test 1`] = `
Array [
  Object {
    "second": 10,
    "text": "hello",
  },
  Object {
    "second": 10,
    "text": "world",
  },
  Object {
    "second": 8,
    "text": "this",
  },
  Object {
    "second": 3,
    "text": "is",
  },
  Object {
    "second": 15,
    "text": "kakaopay",
  },
  Object {
    "second": 5,
    "text": "we",
  },
  Object {
    "second": 5,
    "text": "are",
  },
  Object {
    "second": 15,
    "text": "kakaopay",
  },
  Object {
    "second": 15,
    "text": "frontend",
  },
  Object {
    "second": 20,
    "text": "developers",
  },
  Object {
    "second": 15,
    "text": "join",
  },
  Object {
    "second": 10,
    "text": "us",
  },
]
`;
